### ISSUE TITLE
PEAR.Functions.ValidDefaultValue.NotAtEnd fails on Nullable in PHP 7.1

### ISSUE THREAD

**Issue by TomasVotruba** [squizlabs/PHP_CodeSniffer#1261]
## This is ok

![ok](https://cloud.githubusercontent.com/assets/924196/21745177/7e7d91f8-d525-11e6-9cea-b9d348ce7051.png)

## But when changed to 

![fail](https://cloud.githubusercontent.com/assets/924196/21745180/8842b8bc-d525-11e6-81e7-cf0e0e5a93c1.png)

## It fails like this

![2fail](https://cloud.githubusercontent.com/assets/924196/21745182/92254494-d525-11e6-9888-9cbba42fa2c3.png)

**Comment by gsherwood** [comment]
I think this is working as intended. As an example:
```php
function foo(string $foo, array $bar = [], ?string $baz) {}
foo('hi', [], 'baz'); // valid
foo('hi', [], null); // valid
foo('hi', []); // invalid
```
In your example code, `Response $response = null` makes the argument optional, but `?Response $response` does not. So you can't leave off that 3rd argument as it has no default value, but you can leave off the second argument because it defaults to an empty array.

So from my understanding, your two code examples mean two different things. But please let me know if I've made a mistake - I haven't used nullable types outside of testing.

**Comment by TomasVotruba** [comment]
Well, null is typically used to omit an argument.

Otherwise there would be no default value.

**Comment by gsherwood** [comment]
> Well, null is typically used to omit an argument.

Yes, it absolutely can be. But this sniff is trying to enforce a rule that arguments with default values come at the end of the function declaration. I assume this is called for in standards such as PEAR and PSR2 to ensure that arguments can be omitted entirely from function calls, shortening the length of the function call itself.

Obviously, both standards were written before nullable types were introduced and so maybe they would have a different take on function declarations if they were written today. But I think this sniff accurately reflects the language and intention of those standards because nullable arguments don't actually have default values.

From the original [RFC](https://wiki.php.net/rfc/nullable_types):
> Parameters with a nullable type do not have a default value. If omitted the value does not default to null and will result in an error

And from [PSR2 section 4.4](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-2-coding-style-guide.md#44-method-arguments):
> Method arguments with default values MUST go at the end of the argument list.

If you disagree, you could always ask on the PSR2 mailing list and see if you can get one of the PSR2 authors to comment.

**Comment by TomasVotruba** [comment]
The issue is: those two code snippets are equal. They should both fail or both pass.

**Comment by gsherwood** [comment]
They are not equal.

In the first code snippet, the third argument is optional, so you can call the function without passing any value for it. In the second code snippet, the argument is not optional; you must pass an object or null when calling the function.

See my first comment for code examples and another explanation.

**Comment by TomasVotruba** [comment]
Oh, I didn't know that. You are right: https://3v4l.org/L95Ea

I'm sory, it works as expected then.

**Comment by gsherwood** [comment]
To be honest, I didn't know until I started looking into it today either. Glad it all makes sense now though. Thanks for doing that testing.
